packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -2897,41 +2898,52 @@ let HooksDispatcherOnMountInDEV: Dispatcher | null = null;
 let HooksDispatcherOnMountWithHookTypesInDEV: Dispatcher | null = null;
 let HooksDispatcherOnUpdateInDEV: Dispatcher | null = null;
 let HooksDispatcherOnRerenderInDEV: Dispatcher | null = null;
-let InvalidNestedHooksDispatcherOnMountInDEV: Dispatcher | null = null;
-let InvalidNestedHooksDispatcherOnUpdateInDEV: Dispatcher | null = null;
-let InvalidNestedHooksDispatcherOnRerenderInDEV: Dispatcher | null = null;
 
 if (__DEV__) {
   const warnInvalidContextAccess = () => {
-    console.error(
-      'Context can only be read while React is rendering. ' +
-        'In classes, you can read it in the render method or getDerivedStateFromProps. ' +
-        'In function components, you can read it directly in the function body, but not ' +
-        'inside Hooks like useReducer() or useMemo().',
-    );
+    if (isExecutingHookDev) {
+      console.error(
+        'Context can only be read while React is rendering. ' +
+          'In classes, you can read it in the render method or getDerivedStateFromProps. ' +
+          'In function components, you can read it directly in the function body, but not ' +
+          'inside Hooks like useReducer() or useMemo().',
+      );
+    }
   };
 
   const warnInvalidHookAccess = () => {
-    console.error(
-      'Do not call Hooks inside useEffect(...), useMemo(...), or other built-in Hooks. ' +
-        'You can only call Hooks at the top level of your React function. ' +
-        'For more information, see ' +
-        'https://reactjs.org/link/rules-of-hooks',
-    );
+    if (isExecutingHookDev) {
+      console.error(
+        'Do not call Hooks inside useEffect(...), useMemo(...), or other built-in Hooks. ' +
+          'You can only call Hooks at the top level of your React function. ' +
+          'For more information, see ' +
+          'https://reactjs.org/link/rules-of-hooks',
+      );
+    }
+  };
+
+  ContextOnlyDispatcher.readContext = function <T>(
+    context: ReactContext<T>,
+  ): T {
+    warnInvalidContextAccess();
+    return readContext(context);
```

Comment:
> Cannot assign function to `ContextOnlyDispatcher.readContext` because property `readContext` is **_not writable_**.

*One of the ways* to tackle this will be to _refactor/declare_ the `warnInvalidContextAccess` fn such that:
- it'll behave as warn-**If**-InvalidContextAccess (i.e. `warnIfInvalidContextAccess`)
- which takes care of __*if*__ the access is _invalid_ on top of current behaviour
- may use `__DEV__`(?)

and declaring `ContextOnlyDispatcher` as:
```ts
export const ContextOnlyDispatcher: Dispatcher = {
  readContext: function <T>(
    context: ReactContext<T>,
  ): T {
    warnIfInvalidContextAccess();// the above refactored fn
    return readContext(context);
  },
  ...,
}
```